server/src/main/java/org/elasticsearch/tasks/CancellableTasksTracker.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+/**
+ * Tracks items that are associated with cancellable tasks, supporting efficient lookup by task ID and by parent task ID
+ */
+public class CancellableTasksTracker<T> {
+
+    private final T[] empty;
+
+    public CancellableTasksTracker(T[] empty) {
+        assert empty.length == 0;
+        this.empty = empty;
+    }
+
+    private final Map<Long, T> byTaskId = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+    private final Map<TaskId, T[]> byParentTaskId = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
```

Comment:
Does it make sense to store tasks as `List<T>` instead of `T[]` for the clarity sake if we don't do any specific optimizations? We could simplify code by relying on the `List` API for copying and inserting data instead of `Arrays.copyOf` and `Arrays.copyOfRange`

Reply:
I didn't find a better `List<T>`-based implementation myself, but I'm open to ideas. Would you pull this branch, flesh out your suggestion, push it to your fork and then link to it here?

Reply:
@DaveCTurner Pushed it here! https://github.com/arteam/elasticsearch/commit/a2bdcfbd80b37f509586f4aadf7922dab8da909e

Reply:
We can also potentially use `merge` instead of `compute` to make the addition operation more concise https://github.com/elastic/elasticsearch/commit/a50401fa7b5f77b7c638f624b27b89ea619f002c

Reply:
We can also consider using `merge` instead of `concat` for Lists to make the addition operation more concise 
https://github.com/arteam/elasticsearch/commit/a50401fa7b5f77b7c638f624b27b89ea619f002c

Reply:
The benchmarks indicate that a2bcdfb is a few nanoseconds quicker for the 1-child case but quite substantially slower otherwise (I skipped the 30k-parent cases because they add a lot to the runtime for no real benefit):

```
(childCountPerParent)  (parentCount)  Mode  Cnt            Score           Error  Units  Mean overhead
                 1000          30000                     SKIPPED
                 1000           3000  avgt    7  25940199095.000 ± 142227031.895  ns/op        8647 ns     
                 1000            300  avgt    7   2392018166.114 ±  30498397.694  ns/op        7973 ns     
                 1000             30  avgt    7    218739912.786 ±   1249128.187  ns/op        7291 ns     
                 1000              3  avgt    7     22607793.973 ±    111531.112  ns/op        7536 ns     
                  100          30000                     SKIPPED
                  100           3000  avgt    7    418881040.660 ±  20161249.132  ns/op        1396 ns     
                  100            300  avgt    7     35457186.201 ±    417486.108  ns/op        1182 ns     
                  100             30  avgt    7      3033577.940 ±     97236.034  ns/op        1011 ns     
                  100              3  avgt    7       299040.083 ±     18249.359  ns/op         997 ns     
                   10          30000                     SKIPPED
                   10           3000  avgt    7     12848626.628 ±    235736.397  ns/op         428 ns     
                   10            300  avgt    7       855412.478 ±      5594.042  ns/op         285 ns     
                   10             30  avgt    7        82910.096 ±      2885.424  ns/op         276 ns     
                   10              3  avgt    7         7580.669 ±        28.137  ns/op         253 ns     
                    1          30000                     SKIPPED
                    1           3000  avgt    7       302676.158 ±      8294.205  ns/op         101 ns     
                    1            300  avgt    7        29608.966 ±       889.803  ns/op          99 ns     
                    1             30  avgt    7         2935.642 ±        42.769  ns/op          98 ns     
                    1              3  avgt    7          310.498 ±         1.610  ns/op         103 ns     
```

This code is on a fairly hot path, every time a search touches a shard it will register and unregister a cancellable task for instance, a cluster could be handling thousands of searches per second each touching thousands of shards, so I think it's worth dropping to the raw arrays even if it is a bit harder to read.

I'd also be concerned about the additional memory usage and GC workload, we're making and growing an `ArrayList` in each `.stream().filter(item -> item != oldItem).collect(Collectors.toList())`, which adds quite some overhead, although I will admit I haven't measured this.

Reply:
That sounds alright! I think it totally makes sense to use raw arrays in hot paths :+1: